### PR TITLE
initialize dgfip specific backend

### LINKED ISSUE #19: Create new DGFiP backend
Create new Mlang backend:

- Keep current C backend and add another one that's specific to the DGFiP. 
- Write an M++ file containing a single function that only calls call_m once.

### PR DESCRIPTION
**Pull request by jemsab**
closes #19

### CODE REVIEW

File: examples/c_dgfip/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+### Using the generated C file
+
+The main function provided by the generated C file provides one function,
+`m_extracted`. This functions takes a struct as its argument, and returns another
+struct.
+
+he input struct has its own type, `m_input`, whose fields are the different
+input values of the income tax computation. Each value has type `m_value`, which
+is a `double` wrapped with a boolean flag to indicate if the value is undefined.
```
Comment (denismerigoux) [inline examples/c_dgfip/README.md:30]: I don't think this doc is up to date with what you plan to do with the DGFiP backend, I would prefer you remove the text and put a placeholder until you write specific doc for it.

**Review by denismerigoux — COMMENTED** [review]
I have the impression that is is a draft that duplicates the C backend, I'll switch it to draft until this actually contains the specific DGFiP backend.

**Comment by jemsab** [comment]
Now that I know how to get CI working, it would be probably be a good time to implement tests in the main makefile for the DGFiP backend.

File: examples/dgfip_c/backend_tests/fuzz_harness.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
+#include "ir_tests.h"
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/fuzz_harness.c:1]: I think you can remove this file, this is not a priority for the new backend.

File: examples/dgfip_c/backend_tests/fuzz_inputs/length_of_private_tests_2018 [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis pretium, risus eu porttitor tempor, orci nisi laoreet urna, vel mattis ex tellus eget urna. Donec eu vulputate risus, vitae convallis turpis. In augue nisl, pretium et odio ac, bibendum hendrerit ex. Curabitur vel eleifend felis. Integer vel odio accumsan, consectetur est at, venenatis enim. Curabitur tempus dapibus mauris, nec sollicitudin nisi condimentum vel. Vivamus non condimentum mi, et egestas tellus. Maecenas mollis, libero a tincidunt gravida, lacus nunc porttitor odio, et mollis elit magna sit amet nunc. Duis iaculis vestibulum diam, id sodales augue commodo nec.
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/fuzz_inputs/length_of_private_tests_2018:1]: Idem, can remove

File: examples/dgfip_c/backend_tests/fuzz_inputs/length_of_private_tests_2019 [hunk 4]
```diff
@@ -0,0 +1,19 @@
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla tempus euismod ligula, sit amet blandit elit imperdiet et. Vestibulum sit amet turpis hendrerit, accumsan odio at, aliquet odio. Integer quis eros eu augue feugiat sodales. Nulla ut suscipit nisi. Nulla facilisi. Duis tincidunt auctor ante quis pellentesque. Curabitur ornare luctus convallis. In ex risus, vehicula vitae diam vitae, commodo blandit libero.
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/fuzz_inputs/length_of_private_tests_2019:1]: Idem, can remove

File: examples/dgfip_c/backend_tests/fuzz_inputs/length_of_random_tests_2018 [hunk 5]
```diff
@@ -0,0 +1,17 @@
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus sit amet accumsan arcu. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Aenean rutrum ornare molestie. Suspendisse non molestie turpis, sed mattis velit. Etiam dignissim interdum mattis. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. In quam ipsum, facilisis at aliquet eget, sollicitudin a orci. Aenean sed lorem ut felis faucibus aliquet in sit amet ipsum. Quisque ac ex urna. Phasellus hendrerit congue nisl.
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/fuzz_inputs/length_of_random_tests_2018:1]: Idem, can remove

File: examples/dgfip_c/backend_tests/fuzz_inputs/length_of_simulateur_simplifie_2018 [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam condimentum lectus mauris, vel suscipit ex finibus non. In posuere molestie sem vitae faucibus. Nam sem est, vestibulum id tempus id, lacinia eu purus. Duis elementum purus quis odio dapibus aliquam vel sed sem. Phasellus scelerisque justo ex, a dapibus sapien finibus id. Phasellus accumsan orci eget purus finibus congue. Quisque mauris odio, suscipit sed tincidunt vel, lobortis at sapien. Integer posuere, purus vel dignissim porta, nisl erat bibendum mauris, a faucibus massa ex eu purus. In nec diam nec ipsum tincidunt mollis. Sed ut porta orci, id egestas est.
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/fuzz_inputs/length_of_simulateur_simplifie_2018:1]: Idem, can remove

File: examples/dgfip_c/backend_tests/perf_harness.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
+#include "ir_tests.h"
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/perf_harness.c:1]: I think you can also remove that file

File: examples/dgfip_c/backend_tests/sanitize_crash_names.sh [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#! /bin/bash
```
Comment (denismerigoux) [inline examples/dgfip_c/backend_tests/sanitize_crash_names.sh:1]: Remove, used for fuzzing

File: examples/dgfip_c/run_all_ins_selected_outs_2018.c [hunk 9]
```diff
@@ -0,0 +1,32 @@
+#include "ir_all_ins_selected_outs_2018.h"
+
```
Comment (denismerigoux) [inline examples/dgfip_c/run_all_ins_selected_outs_2018.c:2]: I think you can remove? This is just a toy example program

File: examples/dgfip_c/run_complex_case_2018.c [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#include "ir_complex_case_2018.h"
```
Comment (denismerigoux) [inline examples/dgfip_c/run_complex_case_2018.c:1]: Idem, can remove

File: examples/dgfip_c/run_simulateur_simplifie_2018.c [hunk 11]
```diff
@@ -0,0 +1,32 @@
+#include "ir_simulateur_simplifie_2018.h"
```
Comment (denismerigoux) [inline examples/dgfip_c/run_simulateur_simplifie_2018.c:1]: Idem, can remove

File: examples/dgfip_c/Makefile [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+YEAR?=2019
+
+ifeq ($(YEAR), 2018)
+	SOURCE_FILES?=$(SOURCE_DIR_2018)
+	MPP_FILE?=../../mpp_specs/2018_6_7.mpp
+	TESTS_DIR?=../../tests/2018/fuzzing/
+else ifeq ($(YEAR), 2019)
+	SOURCE_FILES?=$(SOURCE_DIR_2019)
+	MPP_FILE?=../../mpp_specs/2019_8_0.mpp
```
Comment (denismerigoux) [inline examples/dgfip_c/Makefile:19]: This should be changed to `dgfip_base.m` right ?

File: mpp_specs/dgfip_base.mpp [hunk 13]
```diff
@@ -0,0 +1 @@
+outputs <- call_m()
```
Comment (denismerigoux) [inline mpp_specs/dgfip_base.mpp:1]: You're supposed to put that in a function (see other M++ files)

File: src/mlang/driver.ml [hunk 14]
```diff
@@ -155,6 +155,13 @@ let driver (files : string list) (debug : bool) (var_info_debug : string list) (
             Bir_to_c.generate_c_program combined_program function_spec !Cli.output_file;
             Cli.debug_print "Result written to %s" !Cli.output_file
           end
+          else if String.lowercase_ascii backend = "dgfip_c" then begin
+            Cli.debug_print "Compiling the codebase to DGFiP C...";
```
Comment (denismerigoux) [inline src/mlang/driver.ml:159]: You should also modify the usage message related to the backend option in `cli.ml`.

**Comment by jemsab** [comment]
@denismerigoux CI is failing because of the calculation tests. As you know in order to give the right results we have to use the functions in the M++ specs.  How do you want to proceed? Can this branched be merged as is with incorrect results?

**Comment by denismerigoux** [comment]
Hum the problem is that if you want to test the DGFiP backend in real conditions you need the INTER files and the AUTOTEST. So what we really should do is setup the INTER files and autotest on Raphaël's machine, and set up the correct CI pipeline there. But for this PR we can leave this out, and simply remove `test_dgfip_backend` from the current CI pipeline.

File: .gitlab-ci.yml [hunk 15]
```diff
@@ -32,6 +32,11 @@ test_c_backend:
   script:
     - make test_c_backend
 
+test_dgfip_c_backend:
```
Comment (denismerigoux) [inline .gitlab-ci.yml:35]: You can remove that for now

**Review by denismerigoux — CHANGES_REQUESTED** [review]

**Review by denismerigoux — APPROVED** [review]

**Pull request merged**